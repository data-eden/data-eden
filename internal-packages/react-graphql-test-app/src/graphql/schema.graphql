enum Breed {
  SHEPARD
  BULLDOG
  POODLE
  GERMAN_SHEPHERD
  LABRADOR_RETRIEVER
  GOLDEN_RETRIEVER
}

union Owner = Person | Company

type Adoption {
  id: ID!
  pet: Pet
  isAdopted: Boolean!
}

type Person {
  id: ID!
  name: String!
  car: Car!
  pets: [Pet!]!
}

input PersonInput {
  name: String!
}

type Car {
  id: ID!
  make: String!
  model: String!
  owner: Owner!
}

type Company {
  id: ID!
  name: String!
}

input CarInput {
  make: String
  model: String
}

type Pet {
  id: ID!
  name: String!
  owner: Person!
  breed: Breed!
}

type PetsForAdoption {
  id: ID!
  pets: [Pet!]!
}

input UpdatePetInput {
  name: String!
}

input CreatePetInput {
  name: String!
  personId: ID!
}

input RemovePetInput {
  id: ID!
  personId: ID!
}

type Query {
  person(id: ID!): Person!
  car(id: ID!): Car!
  people: [Person!]!
  petsForAdoption: PetsForAdoption!
}

type Mutation {
  updatePerson(personId: ID!, input: PersonInput!): Person!
  updateCar(carId: ID!, input: CarInput!): Car!
  updatePet(petId: ID!, input: UpdatePetInput!): Pet!
  createPet(input: CreatePetInput!): Pet!
  removePet(id: ID!): [Pet!]!
}
