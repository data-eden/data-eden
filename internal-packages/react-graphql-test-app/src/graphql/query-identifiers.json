{
  "8f6f79e36cbc21beda49d817494f6c85e302231458c543d030d1e8af901c7f36": "query Car($id: ID!) { car(id: $id) { __typename id make model owner { __typename id } } }",
  "2b0fc416adc23e2c7cc9761a868714cb5662650a21e9fe14bfa358d98c68b82c": "fragment PersonFields on Person { __typename id name } query Person($id: ID!) { person(id: $id) { __typename car { __typename id make model } pets { __typename id name owner { __typename ...PersonFields } } ...PersonFields } }",
  "d3e6d41ef1b45d4ebe7c412d84e569b2ea8af1df6676bf86af0e2504fb0e46ab": "mutation CreatePet($input: CreatePetInput!) { createPet(input: $input) { __typename id name owner { __typename id pets { __typename id name owner { __typename id name pets { __typename id name owner { __typename id name pets { __typename id name owner { __typename id name } } } } } } } } }",
  "402b683c9e489fa1c769d7cd5815b644acdfb36179d8a9cfa7055e0f128698dd": "mutation RemovePet($id: ID!) { removePet(id: $id) { __typename id name owner { __typename id pets { __typename id name owner { __typename id name pets { __typename id name owner { __typename id name } } } } } } }",
  "38a98bbd5e7b6463f8dda2de92b3b99c531ca7d53af036685dbc366e12ffc6da": "mutation UpdateCar($carId: ID!, $input: CarInput!) { updateCar(carId: $carId, input: $input) { __typename id make model } }",
  "1eb1de91b9d83bd2e44c39aa32ebc7fa50cf65176be52f403c5abef75f6defef": "mutation UpdatePerson($input: PersonInput!, $personId: ID!) { updatePerson(personId: $personId, input: $input) { __typename id name } }",
  "af5c6749a82a0b93e9bce39adf9627e4fb5fd249da1010f72650417b155a5f65": "mutation UpdatePet($input: UpdatePetInput!, $petId: ID!) { updatePet(petId: $petId, input: $input) { __typename id name } }",
  "4eb0d9d1a2901538f228ca5652d69bf63c60f0864fdc0450e3c1f5d30968bf10": "fragment CarFields on Car { id make model } fragment PersonFieldsShared on Person { __typename id name } fragment PetFields on Pet { __typename id name owner { ...PersonFieldsShared } } query Person($id: ID!) { person(id: $id) { __typename car { __typename ...CarFields } pets { __typename ...PetFields } ...PersonFieldsShared } }",
  "c14630fad5e9dcdd6b37a7a7d1d836c1a7eea9950778e188d7705b1d119540ac": "query PetsForAdoption { petsForAdoption { __typename id pets { __typename id name owner { __typename id name } } } }",
  "fb1de7b0677530702610754ef9906528b8f6d42dfc76d34f9acfd51850c4c52e": "query Car($id: ID!) { car(id: $id) { __typename id make model owner { __typename ... on Person { __typename id } } } }"
}
