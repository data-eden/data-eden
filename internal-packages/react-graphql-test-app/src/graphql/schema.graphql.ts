export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export enum Breed {
  Shepard = 'SHEPARD',
  Bulldog = 'BULLDOG',
  Poodle = 'POODLE',
  GermanShepherd = 'GERMAN_SHEPHERD',
  LabradorRetriever = 'LABRADOR_RETRIEVER',
  GoldenRetriever = 'GOLDEN_RETRIEVER'
}

export type PersonInput = {
  name: Scalars['String'];
};

export type CarInput = {
  make?: InputMaybe<Scalars['String']>;
  model?: InputMaybe<Scalars['String']>;
};

export type UpdatePetInput = {
  name: Scalars['String'];
};

export type CreatePetInput = {
  name: Scalars['String'];
  personId: Scalars['ID'];
};

export type RemovePetInput = {
  id: Scalars['ID'];
  personId: Scalars['ID'];
};
