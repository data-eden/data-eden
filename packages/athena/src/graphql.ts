import type { TypedDocumentNode } from '@graphql-typed-document-node/core';
import type { DefaultVariables } from './types.js';

export function graphql<
  Data extends object = object,
  Variables extends DefaultVariables = DefaultVariables
>(
  literals: TemplateStringsArray,
  ...args: any[]
): TypedDocumentNode<Data, Variables> {
  throw new Error(
    'This function should never actually be evaluated. Please use the babel plugin.'
  );
}
