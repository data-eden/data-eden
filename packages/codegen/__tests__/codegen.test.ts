import { printExecutableGraphQLDocument } from '@graphql-tools/documents';
import type { Project } from 'fixturify-project';
import { visit } from 'graphql';
import * as crypto from 'node:crypto';
import * as fs from 'node:fs/promises';
import { existsSync } from 'node:fs';
import * as path from 'node:path';
import { describe, expect, test } from 'vitest';
import { athenaCodegen } from '../src/index.js';
import {
  createFixtures,
  gqlFilesIgnoreMap,
  gqlFilesMap,
  gqlFilesMapWithReExportedFragments,
  gqlFilesMapWithSharedFragments,
  gqlFilesMapWithSharedFragmentsTransitive,
  gqlFilesMapWithSharedFragmentsUsedAndExported,
  graphqlFilesMap,
} from './utils/project.js';

describe('codegen', () => {
  let project: Project;

  async function read(relativePath: string) {
    return await fs.readFile(path.join(project.baseDir, relativePath), 'utf-8');
  }

  test('with graphql files', async () => {
    project = await createFixtures(graphqlFilesMap);

    await athenaCodegen({
      schemaPath: 'schema.graphql',
      documents: [`${project.baseDir}/**/*.graphql`],
      baseDir: project.baseDir,
      extension: '.graphql.ts',
      disableSchemaTypesGeneration: false,
      production: false,
    });

    expect(await read('queries/__generated/my-chats.graphql.ts'))
      .toMatchInlineSnapshot(`
        "import type * as Types from '../../schema.graphql.js';

        import type { ChatFieldsFragment } from '../../fragments/__generated/chat-fields-fragment.graphql.js';
        import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
        export type ChatsQueryVariables = Types.Exact<{
          userId: Types.Scalars['ID'];
        }>;


        export type ChatsQuery = { __typename: 'Query', myChats: Array<(
            { __typename: 'Chat' }
            & ChatFieldsFragment
          )> };


        export const ChatsDocument = {\\"__meta__\\":{\\"queryId\\":\\"a0cd151991e13b081cc250bf42218547f2c8e340282b90c1f7fe36c332de3416\\",\\"$DEBUG\\":{\\"contents\\":\\"fragment ChatFields on Chat { id messages { content id user { id username } } users { ...UserFields } } fragment UserFields on User { id role username } query chats($userId: ID!) { myChats { __typename ...ChatFields } }\\"}}} as unknown as DocumentNode<ChatsQuery, ChatsQueryVariables>;"
      `);

    expect(await read('query-identifiers.json')).toMatchInlineSnapshot(`
      "{
        \\"d17490e4b9ac1f7c227df3da6e5c5cdc6686b24d7194c0cb1bc29a8189a58f5c\\": \\"fragment UserFields on User { id role username } query findUser($userId: ID!) { user(id: $userId) { __typename ...UserFields } }\\",
        \\"a0cd151991e13b081cc250bf42218547f2c8e340282b90c1f7fe36c332de3416\\": \\"fragment ChatFields on Chat { id messages { content id user { id username } } users { ...UserFields } } fragment UserFields on User { id role username } query chats($userId: ID!) { myChats { __typename ...ChatFields } }\\"
      }"
    `);

    expect(await read('schema.graphql.ts')).toMatchInlineSnapshot(`
      "export type Maybe<T> = T | null;
      export type InputMaybe<T> = Maybe<T>;
      export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
      export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
      export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
      /** All built-in and custom scalars, mapped to their actual values */
      export type Scalars = {
        ID: string;
        String: string;
        Boolean: boolean;
        Int: number;
        Float: number;
        Date: any;
      };

      export enum Role {
        User = 'USER',
        Admin = 'ADMIN'
      }
      "
    `);
  });

  test('with graphql files (production)', async () => {
    project = await createFixtures(graphqlFilesMap);

    await athenaCodegen({
      schemaPath: 'schema.graphql',
      documents: [`${project.baseDir}/**/*.graphql`],
      baseDir: project.baseDir,
      extension: '.graphql.ts',
      disableSchemaTypesGeneration: false,
      production: true,
    });

    expect(await read('queries/__generated/my-chats.graphql.ts'))
      .toMatchInlineSnapshot(`
        "import type * as Types from '../../schema.graphql.js';

        import type { ChatFieldsFragment } from '../../fragments/__generated/chat-fields-fragment.graphql.js';
        import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
        export type ChatsQueryVariables = Types.Exact<{
          userId: Types.Scalars['ID'];
        }>;


        export type ChatsQuery = { __typename: 'Query', myChats: Array<(
            { __typename: 'Chat' }
            & ChatFieldsFragment
          )> };


        export const ChatsDocument = {\\"__meta__\\":{\\"queryId\\":\\"a0cd151991e13b081cc250bf42218547f2c8e340282b90c1f7fe36c332de3416\\"}} as unknown as DocumentNode<ChatsQuery, ChatsQueryVariables>;"
      `);

    expect(await read('query-identifiers.json')).toMatchInlineSnapshot(`
      "{
        \\"d17490e4b9ac1f7c227df3da6e5c5cdc6686b24d7194c0cb1bc29a8189a58f5c\\": \\"fragment UserFields on User { id role username } query findUser($userId: ID!) { user(id: $userId) { __typename ...UserFields } }\\",
        \\"a0cd151991e13b081cc250bf42218547f2c8e340282b90c1f7fe36c332de3416\\": \\"fragment ChatFields on Chat { id messages { content id user { id username } } users { ...UserFields } } fragment UserFields on User { id role username } query chats($userId: ID!) { myChats { __typename ...ChatFields } }\\"
      }"
    `);

    expect(await read('schema.graphql.ts')).toMatchSnapshot();
  });

  test('with custom hash function', async () => {
    project = await createFixtures(graphqlFilesMap);

    await athenaCodegen({
      schemaPath: 'schema.graphql',
      documents: [`${project.baseDir}/**/*.graphql`],
      baseDir: project.baseDir,
      extension: '.graphql.ts',
      disableSchemaTypesGeneration: false,
      production: false,
      hash(document) {
        let operationName = 'Document';

        visit(document, {
          OperationDefinition(node) {
            let name = node.name?.value;

            if (name) {
              operationName = name;
            }
          },
        });

        const hash = crypto
          .createHash('sha256')
          .update(printExecutableGraphQLDocument(document))
          .digest('hex');

        return `${operationName}-${hash}`;
      },
    });

    expect(await read('queries/__generated/my-chats.graphql.ts'))
      .toMatchInlineSnapshot(`
        "import type * as Types from '../../schema.graphql.js';

        import type { ChatFieldsFragment } from '../../fragments/__generated/chat-fields-fragment.graphql.js';
        import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
        export type ChatsQueryVariables = Types.Exact<{
          userId: Types.Scalars['ID'];
        }>;


        export type ChatsQuery = { __typename: 'Query', myChats: Array<(
            { __typename: 'Chat' }
            & ChatFieldsFragment
          )> };


        export const ChatsDocument = {\\"__meta__\\":{\\"queryId\\":\\"chats-a0cd151991e13b081cc250bf42218547f2c8e340282b90c1f7fe36c332de3416\\",\\"$DEBUG\\":{\\"contents\\":\\"fragment ChatFields on Chat { id messages { content id user { id username } } users { ...UserFields } } fragment UserFields on User { id role username } query chats($userId: ID!) { myChats { __typename ...ChatFields } }\\"}}} as unknown as DocumentNode<ChatsQuery, ChatsQueryVariables>;"
      `);

    expect(await read('query-identifiers.json')).toMatchInlineSnapshot(`
      "{
        \\"findUser-d17490e4b9ac1f7c227df3da6e5c5cdc6686b24d7194c0cb1bc29a8189a58f5c\\": \\"fragment UserFields on User { id role username } query findUser($userId: ID!) { user(id: $userId) { __typename ...UserFields } }\\",
        \\"chats-a0cd151991e13b081cc250bf42218547f2c8e340282b90c1f7fe36c332de3416\\": \\"fragment ChatFields on Chat { id messages { content id user { id username } } users { ...UserFields } } fragment UserFields on User { id role username } query chats($userId: ID!) { myChats { __typename ...ChatFields } }\\"
      }"
    `);

    expect(await read('schema.graphql.ts')).toMatchSnapshot();

    project.dispose();
  });

  test('with gql tags', async () => {
    project = await createFixtures(gqlFilesMap);

    await athenaCodegen({
      schemaPath: 'schema.graphql',
      documents: [`${project.baseDir}/**/*.tsx`],
      baseDir: project.baseDir,
      extension: '.graphql.ts',
      disableSchemaTypesGeneration: false,
      production: false,
    });

    expect(await read('schema.graphql.ts')).toMatchSnapshot();

    expect(await read('__generated/user.graphql.ts')).toMatchInlineSnapshot(`
      "import type * as Types from '../schema.graphql.js';

      import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
      export type UserFieldsFragment = { __typename: 'User', id: string, username: string, role: Types.Role };

      export type FindUserQueryVariables = Types.Exact<{
        userId: Types.Scalars['ID'];
      }>;


      export type FindUserQuery = { __typename: 'Query', user?: (
          { __typename: 'User' }
          & UserFieldsFragment
        ) | null };


      export const FindUserDocument = {\\"__meta__\\":{\\"queryId\\":\\"d17490e4b9ac1f7c227df3da6e5c5cdc6686b24d7194c0cb1bc29a8189a58f5c\\",\\"$DEBUG\\":{\\"contents\\":\\"fragment UserFields on User { id role username } query findUser($userId: ID!) { user(id: $userId) { __typename ...UserFields } }\\"}}} as unknown as DocumentNode<FindUserQuery, FindUserQueryVariables>;"
    `);
  });

  test('with gql tags (production)', async () => {
    project = await createFixtures(gqlFilesMap);

    await athenaCodegen({
      schemaPath: 'schema.graphql',
      documents: [`${project.baseDir}/**/*.tsx`],
      baseDir: project.baseDir,
      extension: '.graphql.ts',
      disableSchemaTypesGeneration: false,
      production: true,
    });

    expect(await read('schema.graphql.ts')).toMatchSnapshot();

    expect(await read('__generated/user.graphql.ts')).toMatchInlineSnapshot(`
      "import type * as Types from '../schema.graphql.js';

      import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
      export type UserFieldsFragment = { __typename: 'User', id: string, username: string, role: Types.Role };

      export type FindUserQueryVariables = Types.Exact<{
        userId: Types.Scalars['ID'];
      }>;


      export type FindUserQuery = { __typename: 'Query', user?: (
          { __typename: 'User' }
          & UserFieldsFragment
        ) | null };


      export const FindUserDocument = {\\"__meta__\\":{\\"queryId\\":\\"d17490e4b9ac1f7c227df3da6e5c5cdc6686b24d7194c0cb1bc29a8189a58f5c\\"}} as unknown as DocumentNode<FindUserQuery, FindUserQueryVariables>;"
    `);
  });

  test('should not parse tags that do not come from @data-eden/codegen', async () => {
    project = await createFixtures(gqlFilesIgnoreMap);

    await athenaCodegen({
      schemaPath: 'schema.graphql',
      documents: [`${project.baseDir}/**/*.tsx`],
      baseDir: project.baseDir,
      extension: '.graphql.ts',
      disableSchemaTypesGeneration: false,
      production: false,
    });

    expect(
      existsSync(
        path.join(project.baseDir, '__generated', 'invalid.graphql.ts')
      )
    ).toBeFalsy();
  });

  test('gql tags with shared fragments', async () => {
    project = await createFixtures(gqlFilesMapWithSharedFragments);

    await athenaCodegen({
      schemaPath: 'schema.graphql',
      documents: [`${project.baseDir}/**/*.tsx`],
      baseDir: project.baseDir,
      extension: '.graphql.ts',
      disableSchemaTypesGeneration: false,
      production: false,
    });

    expect(await read('schema.graphql.ts')).toMatchSnapshot();

    expect(await read('__generated/UserView.graphql.ts'))
      .toMatchInlineSnapshot(`
        "import type * as Types from '../schema.graphql.js';

        import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
        export type UserFieldsFragment = { __typename: 'User', id: string, username: string, email: string, role: Types.Role };
        "
      `);

    expect(await read('__generated/User.graphql.ts')).toMatchInlineSnapshot(`
      "import type * as Types from '../schema.graphql.js';

      import type { UserFieldsFragment } from './UserView.graphql.js';
      import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
      export type FindUserQueryVariables = Types.Exact<{
        userId: Types.Scalars['ID'];
      }>;


      export type FindUserQuery = { __typename: 'Query', user?: (
          { __typename: 'User' }
          & UserFieldsFragment
        ) | null };


      export const FindUserDocument = {\\"__meta__\\":{\\"queryId\\":\\"2a3590156c3f32f68e767b1f7bb03462e515e373190ef90bfea2f9a9cbd82ddb\\",\\"$DEBUG\\":{\\"contents\\":\\"fragment UserFields on User { email id role username } query findUser($userId: ID!) { user(id: $userId) { __typename ...UserFields } }\\"}}} as unknown as DocumentNode<FindUserQuery, FindUserQueryVariables>;"
    `);
  });

  test('gql tags with shared fragments transitive', async () => {
    project = await createFixtures(gqlFilesMapWithSharedFragmentsTransitive);

    await athenaCodegen({
      schemaPath: 'schema.graphql',
      documents: [`${project.baseDir}/**/*.tsx`],
      baseDir: project.baseDir,
      extension: '.graphql.ts',
      disableSchemaTypesGeneration: false,
      production: false,
    });

    expect(await read('schema.graphql.ts')).toMatchSnapshot();

    expect(await read('__generated/MessageView.graphql.ts'))
      .toMatchInlineSnapshot(`
        "import type * as Types from '../schema.graphql.js';

        import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
        export type MessageFieldsFragment = { __typename: 'ChatMessage', id: string, content: string, time: any };
        "
      `);

    expect(await read('__generated/ChatView.graphql.ts'))
      .toMatchInlineSnapshot(`
        "import type * as Types from '../schema.graphql.js';

        import type { MessageFieldsFragment } from './MessageView.graphql.js';
        import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
        export type ChatFieldsFragment = { __typename: 'Chat', id: string, users: Array<{ __typename: 'User', id: string, username: string }>, messages: Array<(
            { __typename: 'ChatMessage' }
            & MessageFieldsFragment
          )> };
        "
      `);

    expect(await read('__generated/UserView.graphql.ts'))
      .toMatchInlineSnapshot(`
        "import type * as Types from '../schema.graphql.js';

        import type { ChatFieldsFragment } from './ChatView.graphql.js';
        import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
        export type UserFieldsFragment = { __typename: 'User', id: string, username: string, role: Types.Role, chats: Array<(
            { __typename: 'Chat' }
            & ChatFieldsFragment
          ) | null> };
        "
      `);

    expect(await read('__generated/User.graphql.ts')).toMatchInlineSnapshot(`
      "import type * as Types from '../schema.graphql.js';

      import type { UserFieldsFragment } from './UserView.graphql.js';
      import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
      export type FindUserQueryVariables = Types.Exact<{
        userId: Types.Scalars['ID'];
      }>;


      export type FindUserQuery = { __typename: 'Query', user?: (
          { __typename: 'User' }
          & UserFieldsFragment
        ) | null };


      export const FindUserDocument = {\\"__meta__\\":{\\"queryId\\":\\"c2143ab7d20638afd504d251c36c3a62fc1a6f383b93f7bf366ecfba2a3f4e11\\",\\"$DEBUG\\":{\\"contents\\":\\"fragment ChatFields on Chat { id messages { ...MessageFields } users { id username } } fragment MessageFields on ChatMessage { content id time } fragment UserFields on User { chats { ...ChatFields } id role username } query findUser($userId: ID!) { user(id: $userId) { __typename ...UserFields } }\\"}}} as unknown as DocumentNode<FindUserQuery, FindUserQueryVariables>;"
    `);
  });

  // TODO: gql tags with re-exported fragments
  test.skip('gql tags with re-exported fragments', async () => {
    project = await createFixtures(gqlFilesMapWithReExportedFragments);

    await athenaCodegen({
      schemaPath: 'schema.graphql',
      documents: [`${project.baseDir}/**/*.tsx`],
      baseDir: project.baseDir,
      extension: '.graphql.ts',
      disableSchemaTypesGeneration: false,
      production: false,
    });

    expect(await read('schema.graphql.ts')).toMatchSnapshot();

    expect(
      await read('__generated/ChatView.graphql.ts')
    ).toMatchInlineSnapshot();
    expect(
      await read('__generated/UserView.graphql.ts')
    ).toMatchInlineSnapshot();
  });

  test('gql tags with fragments locally used + exported', async () => {
    project = await createFixtures(
      gqlFilesMapWithSharedFragmentsUsedAndExported
    );

    await athenaCodegen({
      schemaPath: 'schema.graphql',
      documents: [`${project.baseDir}/**/*.tsx`],
      baseDir: project.baseDir,
      extension: '.graphql.ts',
      disableSchemaTypesGeneration: false,
      production: false,
    });

    expect(await read('schema.graphql.ts')).toMatchSnapshot();

    expect(await read('__generated/ChatView.graphql.ts'))
      .toMatchInlineSnapshot(`
      "import type * as Types from '../schema.graphql.js';

      import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
      export type ChatFieldsFragment = { __typename: 'Chat', id: string, users: Array<{ __typename: 'User', id: string, username: string }> };

      export type FindMyChatsQueryVariables = Types.Exact<{ [key: string]: never; }>;


      export type FindMyChatsQuery = { __typename: 'Query', myChats: Array<(
          { __typename: 'Chat' }
          & ChatFieldsFragment
        )> };


      export const FindMyChatsDocument = {\\"__meta__\\":{\\"queryId\\":\\"9a1cceea656312633d2f6890193132c1776249e695be24f6e5e2e3da8259a162\\",\\"$DEBUG\\":{\\"contents\\":\\"fragment ChatFields on Chat { id users { id username } } query findMyChats { myChats { __typename ...ChatFields } }\\"}}} as unknown as DocumentNode<FindMyChatsQuery, FindMyChatsQueryVariables>;"
    `);
    expect(await read('__generated/User.graphql.ts')).toMatchInlineSnapshot(`
      "import type * as Types from '../schema.graphql.js';

      import type { ChatFieldsFragment } from './ChatView.graphql.js';
      import type { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
      export type FindUserQueryVariables = Types.Exact<{
        userId: Types.Scalars['ID'];
      }>;


      export type FindUserQuery = { __typename: 'Query', user?: { __typename: 'User', chats: Array<(
            { __typename: 'Chat' }
            & ChatFieldsFragment
          ) | null> } | null };


      export const FindUserDocument = {\\"__meta__\\":{\\"queryId\\":\\"fbb63dc6cd9b42699045d88fb71d7c15042668eec947a17f21acb655c258d031\\",\\"$DEBUG\\":{\\"contents\\":\\"fragment ChatFields on Chat { id users { id username } } query findUser($userId: ID!) { user(id: $userId) { __typename chats { __typename ...ChatFields } } }\\"}}} as unknown as DocumentNode<FindUserQuery, FindUserQueryVariables>;"
    `);
  });

  // TODO: import fragment with aliasing
  //
  // TODO: import fragment from foo.graphql inside a .tsx
  // TODO: import fragment from foo.tsx inside a .graphql
});
