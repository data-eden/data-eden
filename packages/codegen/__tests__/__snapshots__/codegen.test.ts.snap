// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codegen > gql tags with fragments locally used + exported 1`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Date: any;
};

export enum Role {
  User = 'USER',
  Admin = 'ADMIN'
}
"
`;

exports[`codegen > gql tags with shared fragments 1`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Date: any;
};

export enum Role {
  User = 'USER',
  Admin = 'ADMIN'
}
"
`;

exports[`codegen > gql tags with shared fragments coming from aliased modules 1`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Date: any;
};

export enum Role {
  User = 'USER',
  Admin = 'ADMIN'
}
"
`;

exports[`codegen > gql tags with shared fragments transitive 1`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Date: any;
};

export enum Role {
  User = 'USER',
  Admin = 'ADMIN'
}
"
`;

exports[`codegen > with custom hash function 3`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Date: any;
};

export enum Role {
  User = 'USER',
  Admin = 'ADMIN'
}
"
`;

exports[`codegen > with gql tags (production) 1`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Date: any;
};

export enum Role {
  User = 'USER',
  Admin = 'ADMIN'
}
"
`;

exports[`codegen > with gql tags 1`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Date: any;
};

export enum Role {
  User = 'USER',
  Admin = 'ADMIN'
}
"
`;

exports[`codegen > with graphql files (production) 3`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Date: any;
};

export enum Role {
  User = 'USER',
  Admin = 'ADMIN'
}
"
`;
